util/sym-mode.go

```diff
@@ -0,0 +1,262 @@
+// Mgmt
+// Copyright (C) 2013-2019+ James Shubin and the project contributors
+// Written by James Shubin <james@shubin.ca> and the project contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package util
+
+import (
+	"os"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	ModeUser  uint32 = 64
+	ModeGroup uint32 = 8
+	ModeOther uint32 = 1
+
+	ModeRead  uint32 = 4
+	ModeWrite uint32 = 2
+	ModeExec  uint32 = 1
+
+	ModeSetU   uint32 = 4
+	ModeSetG   uint32 = 2
+	ModeSticky uint32 = 1
+)
+
+func modeIsValidWho(who string) bool {
+	for _, w := range []string{"u", "g", "o"} {
+		who = strings.ReplaceAll(who, w, "")
+	}
+	return len(who) == 0
+}
+
+func modeIsValidWhat(what string) bool {
+	for _, w := range []string{"w", "r", "x", "s", "t"} {
+		what = strings.ReplaceAll(what, w, "")
+	}
+	return len(what) == 0
+}
+
+// SymbolicMode is a function that parses a string of symbolic file mode and
+// applies it to the from FileMode.
+//
+// SymbolicFile mode is expected to be a list of who (user, group, other) then
+// the operation (=, +, -) then the change (read, write, execute, setuid,
+// setgid, sticky).
+//
+// Ex. ug=rw
+func SymbolicMode(mode string, from os.FileMode) (os.FileMode, error) {
+	// parse the who
+	// var i int
+	if strings.Contains(mode, "=") && !strings.ContainsAny(mode, "+-") {
```

Comment:
Why isn't this `mode == "="` ?

Reply:
My strategy was to check that the mode string (ex `u=rw`) containes a `=` anywhere and no `+/-` which is the bare minimum check to say it is an assignment.

Reply:
Oh, okay. I'm okay with this, can you please add a short, clear sentence comment explaining that logic at the top of the `if` please? I might just be out of it today, but since this is sort of untraditional for parsing style, a quick reminder will help! =D